File: clothing-erp-project/backend/src/main/java/com/clothing/erp/entity/Customer.java
package com.clothing.erp.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "customer")
public class Customer {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shop_id", nullable = false)
    private Shop shop;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(length = 20)
    private String phone;

    @Column(length = 500)
    private String remark;

    @Column(name = "total_debt", nullable = false, precision = 12, scale = 2)
    private BigDecimal totalDebt = BigDecimal.ZERO;

    @Column(name = "last_purchase_at")
    private LocalDateTime lastPurchaseAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "customer", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<CustomerPrice> specialPrices;

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        this.createdAt = now;
        this.updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}

File: clothing-erp-project/frontend/lib/models/customer.dart
class Customer {
  final String? id;
  final String? name;
  final String? phone;
  final String? remark;
  final double? totalDebt;
  final String? lastPurchaseAt;

  Customer({
    this.id,
    this.name,
    this.phone,
    this.remark,
    this.totalDebt,
    this.lastPurchaseAt,
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id: json['id'] as String?,
      name: json['name'] as String?,
      phone: json['phone'] as String?,
      remark: json['remark'] as String?,
      totalDebt: (json['totalDebt'] as num?)?.toDouble(),
      lastPurchaseAt: json['lastPurchaseAt'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'phone': phone,
      'remark': remark,
      'totalDebt': totalDebt,
      'lastPurchaseAt': lastPurchaseAt,
    };
  }

  Customer copyWith({
    String? id,
    String? name,
    String? phone,
    String? remark,
    double? totalDebt,
    String? lastPurchaseAt,
  }) {
    return Customer(
      id: id ?? this.id,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      remark: remark ?? this.remark,
      totalDebt: totalDebt ?? this.totalDebt,
      lastPurchaseAt: lastPurchaseAt ?? this.lastPurchaseAt,
    );
  }
}
